src/v/compat/tests/compat_test.cc

```diff
@@ -20,12 +22,143 @@
 #include <boost/test/unit_test.hpp>
 
 namespace {
+// NOLINTNEXTLINE(misc-no-recursion)
+std::optional<bool> perturb(json::Value& value, std::string& path) {
```

Comment:
This is super cool but I think there should some sort of non determinism built into this..meaning we should choose to skip perturbing the input with some probability. I think so because the default equality comparison short circuits the remaining checks at first mismatch.. Example.

> class Foo {
>   int a,
>   complex_type b.
> }

If we always perturb `a`, we don't find any bugs in `b`? (or am I totally missing something here?).

Reply:
you are right. it always stops at the first change. I think we can add some non-determinism, great idea.

Reply:
i added a follow up ticket to make this improvement: https://github.com/redpanda-data/redpanda/issues/5902